sdk-extensions/tracing-incubator/src/main/java/io/opentelemetry/sdk/extension/incubator/trace/ExecutorServiceSpanProcessor.java

```diff
@@ -0,0 +1,280 @@
+    private final SpanExporter spanExporter;
+    private final ScheduledExecutorService executorService;
+
+    private Worker(
+        SpanExporter spanExporter,
+        long scheduleDelayNanos,
+        int maxExportBatchSize,
+        long exporterTimeoutNanos,
+        BlockingQueue<ReadableSpan> queue,
+        String spanProcessorTypeLabel,
+        String spanProcessorTypeValue,
+        ScheduledExecutorService executorService,
+        AtomicBoolean isShutdown,
+        long workerScheduleIntervalNanos) {
+      Meter meter = GlobalMeterProvider.getMeter("io.opentelemetry.sdk.trace");
+      meter
+          .longValueObserverBuilder("queueSize")
+          .setDescription("The number of spans queued")
+          .setUnit("1")
+          .setUpdater(
+              result ->
+                  result.observe(
+                      queue.size(), Labels.of(spanProcessorTypeLabel, spanProcessorTypeValue)))
+          .build();
+      LongCounter processedSpansCounter =
+          meter
+              .longCounterBuilder("processedSpans")
+              .setUnit("1")
+              .setDescription(
+                  "The number of spans processed by the BatchSpanProcessor. "
+                      + "[dropped=true if they were dropped due to high throughput]")
+              .build();
+      droppedSpans =
+          processedSpansCounter.bind(
+              Labels.of(spanProcessorTypeLabel, spanProcessorTypeValue, "dropped", "true"));
+      BoundLongCounter exportedSpans =
+          processedSpansCounter.bind(
+              Labels.of(spanProcessorTypeLabel, spanProcessorTypeValue, "dropped", "false"));
+      this.isShutdown = isShutdown;
+      this.executorService = executorService;
+      this.spanExporter = spanExporter;
+      this.batch = new ArrayBlockingQueue<>(maxExportBatchSize);
+      this.workerScheduleIntervalNanos = workerScheduleIntervalNanos;
+      this.maxExportBatchSize = maxExportBatchSize;
+      this.workerExporter =
+          new WorkerExporter(
+              spanExporter,
+              executorService,
+              Logger.getLogger(getClass().getName()),
+              exporterTimeoutNanos,
+              exportedSpans,
+              flushRequested,
+              maxExportBatchSize);
+      this.scheduleDelayNanos = scheduleDelayNanos;
+      this.queue = queue;
+      updateNextExportTime();
+    }
+
+    private void updateNextExportTime() {
+      nextExportTime.set(System.nanoTime() + scheduleDelayNanos);
+    }
+
+    @Override
+    public void run() {
+      // nextExportTime is set for the first time in the constructor
+
+      boolean continueWork = true;
+      while (continueWork && !isShutdown.get()) {
+        if (flushRequested.get() != null) {
+          workerExporter.flush(batch, queue);
+        }
+
+        try {
+          ReadableSpan lastElement = queue.peek();
+          if (lastElement != null) {
+            batch.add(lastElement.toSpanData());
+            // drain queue
```

Comment:
this comment doesn't seem accurate. You're not draining the queue, you're just removing the element you just `peek()`ed at, correct? 
Also, in the CPU-optimization work that was done for the main BSP, we definitely found a significant overhead in working with an ArrayBlockingQueue this way. Any thoughts to using the same signalling technique and the jctools queue here?

Reply:
I'll try to plug in jctools queue.

Reply:
all tests pass 👍